borg/remote.py

```diff
@@ -28,6 +29,21 @@
 MAX_INFLIGHT = 100
 
 
+def os_write(fd, data):
+    """os.write wrapper so we do not lose data for partial writes."""
+    # This is happening frequently on cygwin due to its small pipe buffer size of only 64kiB
+    # and also due to its different blocking pipe behaviour compared to Linux/*BSD.
+    # Neither Linux nor *BSD ever do partial writes on blocking pipes, unless interrupted by a
+    # signal, in which case serve() would terminate.
+    amount = len(data)
+    while data:
+        count = os.write(fd, data)
+        data = data[count:]
```

Comment:
I thought about memoryview, but considering these 64kiB writes and sleep waiting is rather inefficient anyway I thought I keep it simple.

Using count == 0 as termination criteria: i don't think this is the correct criteria. Also, even if it would never return 0 if there was something to send, it would need an extra, empty write call to return 0.

I added the sleep to avoid too fast looping there. If it does a partial write, it means the buffer is currently full, so it might be rather pointless to just try again immediately without waiting a bit and give it a chance to empty the buffer.